### PR TITLE
 Allow nodes to have their post_effect_color affected by lighting

### LINKED ISSUE #12071: "Shaded" post effect color
## Problem

Water nodes that have the same `post_effect_color` as their texture color look strange at night when you are inside them. That's because the `post_effect_color` isn't affected by lighting.

day, outside

day, inside

night, outside

night, inside

## Solutions

Add another (yes, I know, another…) node definition field called `post_effect_color_shaded` or something that is affected by lighting.

## Alternatives

Use a darker `post_effect_color` that looks OK in all lighting situations, but always a bit wrong.

### PR DESCRIPTION
**Pull request by grorp**
Closes #12071.

This PR adds a new boolean field called `post_effect_color_shaded` to node definitions. If this field is true, the `post_effect_color` of the node will be affected by lighting. The default value is false, so nothing changes for existing mods.
Applying lighting to `post_effect_color` makes sense because `post_effect_color` is not applied by color multiplication, but by alpha blending. Lighting is done the same way as for the wielditem.

Here are two videos so you can see the effect:
`post_effect_color_shaded = false`: https://share.grecloud.de/without%20post_effect_color_shaded%202.mp4
`post_effect_color_shaded = true`: https://share.grecloud.de/with%20post_effect_color_shaded%202.mp4

The videos were taken with Minetest Game, the Soothing 32 texture pack and `post_effect_color = {r = 40, g = 204, b = 223, a = 100}` for `default:water_source` and `default:water_flowing`.
Notice that in the first video, the scene is much brighter at night when I'm underwater than when I'm not. In contrast, in the second video, the brightness remains relatively constant at night, regardless of whether I'm underwater or not. To me, the second video looks much better.

**Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?**
Maybe "2.1 Rendering/Graphics improvements".

## To do

This PR is a Ready for Review.

## How to test

Verify that the `post_effect_color` of existing nodes still looks the same.

Try adding `post_effect_color_shaded = true` to the definition of an existing node. Verify that the `post_effect_color` of the node is now affected by lighting. This isn't only visible at night, but also when you go into deep, dark water.

### CODE REVIEW

**Comment by Desour** [comment]
I support this feature.

File: doc/lua_api.md [hunk 1]
```diff
@@ -8702,7 +8702,13 @@ Used by `minetest.register_node`.
     -- Only when `paramtype2` supports palettes.
 
     post_effect_color = "#00000000",
-    -- Screen tint if player is inside node, see "ColorSpec"
+    -- Screen tint if a player is inside this node, see `ColorSpec`.
+    -- Isn't applied by color multiplication, but by drawing a rectangle of this
+    -- color over the entire screen. This means that a `post_effect_color` whose
+    -- alpha is 0 has no effect.
```
Comment (Desour) [inline doc/lua_api.md:8708]: Suggestion:
```suggestion
    -- Color is alpha-blended over the screen.
```

**Review by Desour — COMMENTED** [review]
Code looks fine.

Please add a test node to devtest. And maybe activate the feature for devtest basenode water.

Tested with devtest basenode water.
I'm a bit worried how the color *seems* to get more transparent the deeper you go. Maybe this is some issue related to srgb nonlinearity.

**Comment by grorp** [comment]
> Please add a test node to devtest. And maybe activate the feature for devtest basenode water.

Done.
 
> I'm a bit worried how the color _seems_ to get more transparent the deeper you go. Maybe this is some issue related to srgb nonlinearity.

To me, it doesn't seem like the color is becoming more transparent. And it definitely isn't actually becoming more transparent. But it's very well possible that there are problems with linear / non-linear color everywhere in Minetest.

File: games/devtest/mods/testnodes/properties.lua [hunk 2]
```diff
@@ -588,3 +588,36 @@ minetest.register_node("testnodes:drowning_1", {
 	groups = {dig_immediate=3},
 })
 
+-- post_effect_color_shaded
+
+minetest.register_node("testnodes:post_effect_color_shaded_false", {
+	description = S("\"post_effect_color_shaded = false\" Node"),
+
+	drawtype = "allfaces",
+	tiles = {"testnodes_post_effect_color_shaded_false.png"},
```
Comment (Desour) [inline games/devtest/mods/testnodes/properties.lua:597]: Missing textures.

**Comment by Desour** [comment]
Here are two screenshots, to show what I mean with the transparency increasement:
screenshot_20230716_182139
screenshot_20230716_182157
It might also just be an illusion. After all the sky is getting darker.
(Will approve when the devtest nodes are fixed.)

File: games/devtest/mods/testnodes/properties.lua (hunk 2 above)
Comment (grorp) [reply games/devtest/mods/testnodes/properties.lua:597]: Oops, forgot to `git add` them... Fixed.

**Review by Desour — APPROVED** [review]

**Comment by SmallJoker** [comment]
Is there a particular use-case where it would not be helpful to apply the light multiplication? Comparing the two videos, it appears that this change is much better fit to be applied on all nodes with `post_effect_color` due to more natural appearance.

**Comment by Desour** [comment]
Due to how low-alpha colors look with this features (see screenshots above), imo we should not force this option on content creators.

**Comment by grorp** [comment]
There's another case where modders can decide whether they want lighting or not: The `shaded` object property of entities.

**Pull request merged**